### PR TITLE
Add documentation of new -AsHashtable switch for ConvertFrom-Json introduced by PR #5043

### PR DESCRIPTION
**Pull request by bergmeister**
Add documentation of new `-AsHashTable` switch for `ConvertFrom-Json` and document the behaviour in case of duplicate strings. The switch was introduced by [PR 5043](https://github.com/PowerShell/PowerShell/pull/5043) in version 6.0

Version(s) of document impacted
------------------------------
- [ ] Impacts 6 document

Reason(s) for not updating all version of documents
--------------------------------------------------
- [ ] The documented feature was introduced in version 6.0 of PowerShell

### CODE REVIEW

**Comment by sdwheeler** [comment]
@SteveL-MSFT - can you review and approve this?

**Comment by saldana** [comment]
### :white_check_mark: Validation status: passed


File | Status | Preview URL | Details
---- | ------ | ----------- | -------
[reference/6/Microsoft.PowerShell.Utility/ConvertFrom-Json.md](https://github.com/bergmeister/PowerShell-Docs/blob/convertFromJson_asHashtable_switch/reference/6/Microsoft.PowerShell.Utility/ConvertFrom-Json.md) | :white_check_mark:Succeeded | |

For more details, please refer to the [build report](https://opbuildstorageprod.blob.core.windows.net/report/2017%5C11%5C14%5Cbab27324-4c59-a348-1473-9c6d587e4020%5CPullRequest%5C201711142024105256-1858%5Cworkflow_report.html?sv=2015-02-21&sr=b&sig=O7cNdC1fE5s9nvtnzXVV3LCn6%2BR5MLnncx5rtvK7Jgw%3D&st=2017-11-14T21%3A00%3A25Z&se=2017-12-15T21%3A05%3A25Z&sp=r).

**Note:** If you changed an existing file name or deleted a file, broken links in other files to the deleted or renamed file are listed only in the full build report.

**Comment by saldana** [comment]
### :white_check_mark: Validation status: passed


File | Status | Preview URL | Details
---- | ------ | ----------- | -------
[reference/6/Microsoft.PowerShell.Utility/ConvertFrom-Json.md](https://github.com/bergmeister/PowerShell-Docs/blob/convertFromJson_asHashtable_switch/reference/6/Microsoft.PowerShell.Utility/ConvertFrom-Json.md) | :white_check_mark:Succeeded | |

For more details, please refer to the [build report](https://opbuildstorageprod.blob.core.windows.net/report/2017%5C11%5C14%5Cbab27324-4c59-a348-1473-9c6d587e4020%5CPullRequest%5C201711142025166845-1858%5Cworkflow_report.html?sv=2015-02-21&sr=b&sig=G6SqBZQYau%2BDJ1%2BYkv5cujnOT%2F%2BKQQAIEQ%2FG%2FTpqCdk%3D&st=2017-11-14T21%3A36%3A51Z&se=2017-12-15T21%3A41%3A51Z&sp=r).

**Note:** If you changed an existing file name or deleted a file, broken links in other files to the deleted or renamed file are listed only in the full build report.

File: reference/6/Microsoft.PowerShell.Utility/ConvertFrom-Json.md [hunk 1]
```diff
@@ -11,18 +11,19 @@ title:  ConvertFrom-Json
...
-ConvertFrom-Json [-InputObject] <String> [-InformationAction <ActionPreference>]
- [-InformationVariable <String>] [<CommonParameters>]
+ConvertFrom-Json [-InputObject] <String> [-AsHashtable]
+[-InformationAction <ActionPreference>] [-InformationVariable <String>] [<CommonParameters>]
 ```
 
 ## DESCRIPTION
 The **ConvertFrom-Json** cmdlet converts a JavaScript Object Notation (JSON) formatted string to a custom **PSCustomObject** object that has a property for each field in the JSON string.
 JSON is commonly used by web sites to provide a textual representation of objects.
+The JSON standard does not prohibit duplicate keys and in this case the cmdlet would disregard the duplicates and only use the last one. If keys differ only in casing then the `-AsHashtable` switch has to be used.
```
Comment (SteveL-MSFT) [inline reference/6/Microsoft.PowerShell.Utility/ConvertFrom-Json.md]: I would reword this as:

```markdown
The JSON standard does not prohibit usage that is prohibited with a PSCustomObject.
For example, if the JSON string contains duplicate keys, only the last key is used by this cmdlet.
See other examples below.
```

File: reference/6/Microsoft.PowerShell.Utility/ConvertFrom-Json.md [hunk 2]
```diff
@@ -77,6 +78,14 @@ Then it uses the pipeline operator to send the delimited string to the **Convert
 
 The Join operator is required, because the **ConvertFrom-Json** cmdlet expects a single string.
 
+### Example 4: Convert a JSON string to a hash table
```
Comment (SteveL-MSFT) [inline reference/6/Microsoft.PowerShell.Utility/ConvertFrom-Json.md]: I think it would be more clear to a novice user to separate these into two examples

File: reference/6/Microsoft.PowerShell.Utility/ConvertFrom-Json.md [hunk 3]
```diff
@@ -127,6 +136,25 @@ Accept pipeline input: True (ByValue)
...
+### -AsHashtable
+Converts the JSON to a hash table object. This switch was introduced in PowerShell 6.0.
+There are several scenarios where it can overcome some limitations of the `ConvertFrom-Json` cmdlet
+- If the JSON contains a list with keys that only differ in casing. Without the switch, those keys would be seen as identical keys and therefore only the last one would get used.
+- If the JSON contains a key that is an empty string. Without the switch, the cmdlet would throw an error since a `PSCustomObject` does not allow for that but a hash table does. An example use case where this can occurs are `project.lock.json` files.
+- Hash tables can be processed faster for certain data structures.
+ 
+ ```yaml
+ Type: SwitchParameter
+ Parameter Sets: UNNAMED_PARAMETER_SET_1
```
Comment (SteveL-MSFT) [inline reference/6/Microsoft.PowerShell.Utility/ConvertFrom-Json.md]: Since this switch is not part of a parameter set, I think you just delete this line

File: reference/6/Microsoft.PowerShell.Utility/ConvertFrom-Json.md [hunk 4]
```diff
@@ -127,6 +136,25 @@ Accept pipeline input: True (ByValue)
...
+ Aliases:
+ 
+ Required: False
+ Position: Named
```
Comment (SteveL-MSFT) [inline reference/6/Microsoft.PowerShell.Utility/ConvertFrom-Json.md]: Not a positional parameter, so I think you also delete this

File: reference/6/Microsoft.PowerShell.Utility/ConvertFrom-Json.md [hunk 5]
```diff
@@ -127,6 +136,25 @@ Accept pipeline input: True (ByValue)
...
+ Default value: False
+ Accept pipeline input: True (ByValue)
```
Comment (SteveL-MSFT) [inline reference/6/Microsoft.PowerShell.Utility/ConvertFrom-Json.md]: This should be false as the json is accepted by pipeline

File: reference/6/Microsoft.PowerShell.Utility/ConvertFrom-Json.md (hunk 2 above)
Comment (bergmeister) [reply reference/6/Microsoft.PowerShell.Utility/ConvertFrom-Json.md]: I'll remove the 2nd one (key that is empty string) because this is also documented in the switch documentation and is not so useful as an example

File: reference/6/Microsoft.PowerShell.Utility/ConvertFrom-Json.md (hunk 5 above)
Comment (bergmeister) [reply reference/6/Microsoft.PowerShell.Utility/ConvertFrom-Json.md]: This does not make sense to me. Are you sure?

**Comment by saldana** [comment]
### :white_check_mark: Validation status: passed


File | Status | Preview URL | Details
---- | ------ | ----------- | -------
[reference/6/Microsoft.PowerShell.Utility/ConvertFrom-Json.md](https://github.com/bergmeister/PowerShell-Docs/blob/convertFromJson_asHashtable_switch/reference/6/Microsoft.PowerShell.Utility/ConvertFrom-Json.md) | :white_check_mark:Succeeded | |

For more details, please refer to the [build report](https://opbuildstorageprod.blob.core.windows.net/report/2017%5C11%5C15%5Cbab27324-4c59-a348-1473-9c6d587e4020%5CPullRequest%5C201711151918149277-1858%5Cworkflow_report.html?sv=2015-02-21&sr=b&sig=Y4DZlZiMbdgSkBrkCfzganGX%2Bqzz9d7irjNX%2FsLgtp0%3D&st=2017-11-15T19%3A41%3A01Z&se=2017-12-16T19%3A46%3A01Z&sp=r).

**Note:** If you changed an existing file name or deleted a file, broken links in other files to the deleted or renamed file are listed only in the full build report.

File: reference/6/Microsoft.PowerShell.Utility/ConvertFrom-Json.md (hunk 5 above)
Comment (SteveL-MSFT) [reply reference/6/Microsoft.PowerShell.Utility/ConvertFrom-Json.md]: These values should map to the Parameter attributes:
```c#
        [Parameter(Mandatory = true, Position = 0, ValueFromPipeline = true)]
        [AllowEmptyString]
        public string InputObject { get; set; }

        [Parameter()]
        public SwitchParameter AsHashtable { get; set; }
```

Comment (bergmeister) [reply reference/6/Microsoft.PowerShell.Utility/ConvertFrom-Json.md]: OK, I see now. It confused me at first because one can pipe the string to `ConvertFrom-Json -AsHashtable` but with your example I understand now what it actually refers to.

File: reference/6/Microsoft.PowerShell.Utility/ConvertFrom-Json.md [hunk 6]
```diff
@@ -123,10 +134,27 @@ Aliases:
 Required: True
 Position: 1
 Default value: None
-Accept pipeline input: True (ByValue)
+Accept pipeline input: False
```
Comment (SteveL-MSFT) [inline reference/6/Microsoft.PowerShell.Utility/ConvertFrom-Json.md]: You updated the wrong one.  This one should be reverted.  Line 154 below should be `false`

**Comment by saldana** [comment]
### :white_check_mark: Validation status: passed


File | Status | Preview URL | Details
---- | ------ | ----------- | -------
[reference/6/Microsoft.PowerShell.Utility/ConvertFrom-Json.md](https://github.com/bergmeister/PowerShell-Docs/blob/convertFromJson_asHashtable_switch/reference/6/Microsoft.PowerShell.Utility/ConvertFrom-Json.md) | :white_check_mark:Succeeded | |

For more details, please refer to the [build report](https://opbuildstorageprod.blob.core.windows.net/report/2017%5C11%5C15%5Cbab27324-4c59-a348-1473-9c6d587e4020%5CPullRequest%5C201711152014277395-1858%5Cworkflow_report.html?sv=2015-02-21&sr=b&sig=zbyLbQjNwavPlJNTjTGhO1gOzsP3OhjAOy5Nck%2FDKf0%3D&st=2017-11-15T20%3A40%3A43Z&se=2017-12-16T20%3A45%3A43Z&sp=r).

**Note:** If you changed an existing file name or deleted a file, broken links in other files to the deleted or renamed file are listed only in the full build report.

File: reference/6/Microsoft.PowerShell.Utility/ConvertFrom-Json.md (hunk 6 above)
Comment (bergmeister) [reply reference/6/Microsoft.PowerShell.Utility/ConvertFrom-Json.md]: Oops. Sorry, well spotted. Should be fixed now.

**Comment by saldana** [comment]
### :white_check_mark: Validation status: passed


File | Status | Preview URL | Details
---- | ------ | ----------- | -------
[reference/6/Microsoft.PowerShell.Utility/ConvertFrom-Json.md](https://github.com/bergmeister/PowerShell-Docs/blob/convertFromJson_asHashtable_switch/reference/6/Microsoft.PowerShell.Utility/ConvertFrom-Json.md) | :white_check_mark:Succeeded | |

For more details, please refer to the [build report](https://opbuildstorageprod.blob.core.windows.net/report/2017%5C11%5C15%5Cbab27324-4c59-a348-1473-9c6d587e4020%5CPullRequest%5C201711152102076983-1858%5Cworkflow_report.html?sv=2015-02-21&sr=b&sig=vogdU5X9%2Fn833ptOBmHKgsSXwy3K7bsXsKewYoietPI%3D&st=2017-11-15T21%3A26%3A24Z&se=2017-12-16T21%3A31%3A24Z&sp=r).

**Note:** If you changed an existing file name or deleted a file, broken links in other files to the deleted or renamed file are listed only in the full build report.
